### ISSUE TITLE
Attention

### ISSUE THREAD

**Issue by YujunXie** [YujunXie/Papers-of-Crowd-Counting#3]
<h1 align="center">SENet</h1>
<div align="center">
  《Squeeze-and-Excitation Networks》
</div>

### 简介

- Paper: [TPAMI2018](https://arxiv.org/abs/1709.01507)
- Code: [pytorch](https://github.com/moskomule/senet.pytorch) [tf](https://github.com/taki0112/SENet-Tensorflow)
- 解决问题：自动学习不同通道特征的重要程度。
- 主要贡献：
  - 轻量级的通道注意力机制：对特征图的通道信息赋权重，建模通道之间的依赖关系。

### 方法

- 网络结构：

<img src="https://tva1.sinaimg.cn/large/006tNbRwly1g9h23cdc2uj30ia0hi755.jpg" alt="image-20191201112227734" style="zoom:50%;" />

> adaptively recalibrates channel-wise feature responses by explicitly modelling interdependencies between channels.

1. "squeeze"操作

   一个全局平均池化操作：<img src="https://tva1.sinaimg.cn/large/006tNbRwly1g9iouw27amj30gg038q2x.jpg" alt="image-20191202211534889" style="zoom:50%;" />

   将HxWxC纬度的特征图由空间维度压缩，得到1x1xC纬度的全局描述子（原则上也可以采用更复杂的聚合策略）。

2. "excitation"操作

   为了学习通道信息间的**非线性信息和非互斥关系**。使用sigmoid激活函数的gating机制将通道描述子映射为**通道权值**（值0-1）。两个FC层用于**减小模型复杂度**和**增加泛化性**。其中第一个FC层起到降维的作用，降维系数为r是个超参数，然后采用ReLU激活。最后的FC层恢复原始的维度。

   最后生成的1x1xC维的通道激活值与最初的HxWxC维的特征图逐通道点乘，得到具有通道注意力的特征图。

具体pytorch代码如下：

```pytorch
# SE-Module
class SELayer(nn.Module):
    def __init__(self, channel, reduction=16):
        super(SELayer, self).__init__()
        self.avg_pool = nn.AdaptiveAvgPool2d(1)
        self.fc = nn.Sequential(
            nn.Linear(channel, channel // reduction, bias=False),
            nn.ReLU(inplace=True),
            nn.Linear(channel // reduction, channel, bias=False),
            nn.Sigmoid()
        )
    def forward(self, x):
    b, c, _, _ = x.size()
    y = self.avg_pool(x).view(b, c)
    y = self.fc(y).view(b, c, 1, 1)
    return x * y.expand_as(x)
```

其中主要的参数来源于fc层，可以将fc层换成1x1卷积减少通道数目。

### 实验

- 网络深度的影响：SE模块的引入加深了网络深度，但在增加一小部分模型复杂度的情况下带来更大的结果提升。
- 网络结构：residual, non-residual都能带来大的提升。且在MobileNet等移动端网络上也可部署。
- ablation study: reduction ratio(r越大，参数越少)；全局平均池化/最大值池化；激活函数的不同；网络不同位置。。

### 思考:thinking:

:question:

> 卷积过程本身在 Channel 之间也会有一个乘数，这个乘数是不是本就可以建模Channel 之间的信息。

:dart:  take-home-message

- 关于通道注意力模块在网络浅层和深层的不同作用。

> In earlier layers, it excites informative features in a class-agnostic manner, strengthening the shared low-level representations. In later layers, the SE blocks become increasingly specialised, and respond to different inputs in a highly class-specific manner.

- 本质上，SE模块是在channel维度上做attention或者gating操作，这种注意力机制让模型可以更加关注信息量最大的channel特征，而抑制那些不重要的channel特征。另外一点是SE模块是通用的，这意味着其可以嵌入到现有的网络架构中。

:sparkles: 整个模块的可定制性比较强，可以根据任务的需要进行调整改变。

### 参考文献:books:

**Comment by YujunXie** [comment]
<h1 align="center">SKNet</h1>
<div align="center">
  《Selective Kernel Networks》 
</div>

- Paper: [CVPR2019](https://arxiv.org/pdf/1903.06586.pdf)
- Code: [Caffe]((https://github.com/implus/SKNet)

在SENet的基础上进行了改进，主要有两点：
1. 增加一点计算量换精度。比如使用大一点的 kernel，如 5×5 的卷积提升精度。
2. 结合attention操作。

![image-20191216135930209](https://tva1.sinaimg.cn/large/006tNbRwly1g9yix8xp1sj319o0ckdis.jpg)

>上述结构可以分为三个步骤：

Split：输入为 c×h×w 的特征图，和均表示 Group Convolution。这里使用 Group Convolution 以减少计算量。注意，这里两路 Group Convolution 使用的卷积核大小不一致，原因在于 Motivation 中说的第一点，提升精度。

Fuse：通过 Split 操作分成两路之后，再把两路结果进行融合，然后就是一个 Squeeze and Excitation block。

Select：Select 模块把 Squeeze and Excitation block 模块的结果通过两个 softmax 以回归出 Channel 之间的权重信息。然后把这个权重信息乘到和中。这个过程可以看做是一个 soft attention。最好把两路的特征图进行相加得到输出特征图 V。
